File: problems/2616_minimize_the_maximum_difference_of_pairs/solution.py
from typing import List


class Solution:
    def minimizeMax(self, nums: List[int], p: int) -> int:
        nums.sort()
        lo, hi = 0, nums[-1] - nums[0]
        while lo < hi:
            mid = (lo + hi) // 2
            count = 0
            i = 1
            while i < len(nums):
                if nums[i] - nums[i-1] <= mid:
                    count += 1
                    i += 2
                else:
                    i += 1
            if count >= p: hi = mid
            else: lo = mid + 1
        return lo

File: problems/2616_minimize_the_maximum_difference_of_pairs/solution.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>
#include <climits>
#include <queue>
#include <stack>
#include <cmath>
#include <numeric>
#include <set>
#include <map>
using namespace std;

class Solution {
public:
    int minimizeMax(vector<int>& nums, int p) {
        sort(nums.begin(), nums.end());
        int lo = 0, hi = nums.back() - nums[0];
        while (lo < hi) {
            int mid = (lo+hi)/2, cnt = 0;
            for (int i = 1; i < nums.size(); i++) { if (nums[i]-nums[i-1] <= mid) { cnt++; i++; } }
            if (cnt >= p) hi = mid; else lo = mid+1;
        }
        return lo;
    }
};
